cirq-core/cirq/circuits/circuit.py

```diff
@@ -914,6 +914,9 @@ def all_measurement_key_names(self) -> AbstractSet[str]:
     def _measurement_key_names_(self) -> AbstractSet[str]:
         return self.all_measurement_key_names()
 
+    def _control_key_names_(self):
+        return {key for op in self.all_operations() for key in protocols.control_key_names(op)}
```

Comment:
[No change needed] We'll need to be careful about this - the circuit itself isn't controlled, just the gates inside it.

Reply:
Yeah, my initial thought was to have this return only external keys like CircuitOperations. But for diagramming we'll need a list of *all* the control keys so we can draw the creg lines.

That said, I agree it's odd that the Circuit would have different list than the CircuitOperation. Maybe it's best to have them both hide internal keys, and the diagramming tool can just iterate the circuit's ops and aggregate the control key names manually.

Reply:
(Note to self: I think it's better to have the two be consistent, or people will get confused. Probably best to provide an explicit `include_internal` flag in `control_key_names()`).

Reply:
I don't think we should be exposing internal cbits as `control_keys`. This would really make the semantics confusing as you noted. `control_keys` for any entity (`Operation`, `CircuitOperation` or `Circuit`) should mean "The classical bits that need to be True for this entity to run". So internal keys are not really part of this. We can have a separate protocol to get the internal keys but I don't see what purpose it can serve that cannot already be served by `measurement_keys` protocol. 

The diagram can iterate through internal operations to figure out which of the measurement keys we need to draw a creg for.